File: it-agents/skills/code-architect/references/code-examples/rust/template_method.rs
// Template Method — define the skeleton of an algorithm in a base operation,
// deferring some steps to implementors via overridable trait methods.

trait Pipeline {
    // Overridable steps.
    fn load(&self) -> String;
    fn transform(&self, input: String) -> String;

    // Optional hook with a default.
    fn validate(&self, data: &str) -> bool {
        !data.is_empty()
    }

    // The template method fixes the algorithm's structure.
    fn run(&self) -> Option<String> {
        let raw = self.load();
        if !self.validate(&raw) {
            println!("validation failed");
            return None;
        }
        let result = self.transform(raw);
        println!("pipeline produced: {}", result);
        Some(result)
    }
}

struct UpperPipeline;
struct ReversePipeline;

impl Pipeline for UpperPipeline {
    fn load(&self) -> String {
        "data-one".into()
    }
    fn transform(&self, input: String) -> String {
        input.to_uppercase()
    }
}

impl Pipeline for ReversePipeline {
    fn load(&self) -> String {
        "data-two".into()
    }
    fn transform(&self, input: String) -> String {
        input.chars().rev().collect()
    }
}

fn main() {
    UpperPipeline.run();
    ReversePipeline.run();
}

File: skills/code-architect/references/code-examples/go/template_method.go
// Template Method: define the skeleton of an algorithm in a fixed order,
// deferring individual steps to overridable implementations.
package main

import "fmt"

// steps are the customisable parts of the algorithm.
type steps interface {
	acquire() string
	transform(string) string
}

// run is the invariant template: it fixes the order of the steps.
func run(s steps) string {
	raw := s.acquire()
	return s.transform(raw)
}

// upperFlow is one concrete implementation of the steps.
type upperFlow struct{}

func (upperFlow) acquire() string             { return "abc" }
func (upperFlow) transform(in string) string  { return "UP:" + in }

// reverseFlow is another implementation reusing the same template.
type reverseFlow struct{}

func (reverseFlow) acquire() string            { return "xyz" }
func (reverseFlow) transform(in string) string { return "REV:" + in }

func main() {
	fmt.Println(run(upperFlow{}))
	fmt.Println(run(reverseFlow{}))
}
